middleware.go

```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
+
+	return
+}
+
+func (l loggingInterceptor) interceptStream(
+	srv interface{},
+	ss grpc.ServerStream,
+	info *grpc.StreamServerInfo,
+	handler grpc.StreamHandler,
+) (err error) {
+	startTime := time.Now()
+	ctx := ctxlogrus.ToContext(ss.Context(), logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	wrapped := grpc_middleware.WrapServerStream(ss)
+	wrapped.WrappedContext = ctx
+
+	err = handler(srv, wrapped)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l *loggingInterceptor) requestFromContext(ctx context.Context, request *GRPCRequest) *GRPCRequest {
+	if d, ok := ctx.Deadline(); ok {
+		request.Deadline = d.UTC().Format(time.RFC3339Nano)
+	}
 
-			log.WithFields(fields).Info("Completed request")
+	if p, ok := peer.FromContext(ctx); ok && p != nil {
+		u := &url.URL{
+			Scheme: p.Addr.Network(),
+			Host:   p.Addr.String(),
+		}
+		request.PeerAddr = u.String()
+	}
+
+	if md, ok := metadata.FromIncomingContext(ctx); ok && md != nil {
+		request.UserAgent = strings.Join(md.Get("user-agent"), "")
+
+		if trace := md.Get("X-Cloud-Trace-Context"); len(trace) == 1 {
+			ctxlogrus.AddFields(ctx, logrus.Fields{"trace": trace[0]})
+		}
+	}
+
+	ctxlogrus.AddFields(ctx, logrus.Fields{"grpcRequest": request})
+
+	return request
+}
+
+// logStatus adds the gRPC Status to the log context.
+// If the response is an internal server error, log that as an Error
+// returns true if the logging was handled (e.g. internal server error)
+func (l *loggingInterceptor) log(ctx context.Context, err error, method string) {
+	if !l.filterRPC(ctx, method, err) {
+		return
+	}
+
+	if err != nil {
+		st := status.Convert(err)
+		jsonStatus, merr := ProtoJSONMarshalOptions.Marshal(st.Proto())
+		if merr != nil {
+			ctxlogrus.Extract(ctx).WithError(merr).Errorf("error marshalling error status into log")
+		}
+		ctxlogrus.AddFields(ctx, logrus.Fields{
+			"grpcStatus": json.RawMessage(jsonStatus),
 		})
+
+		// if the error was already logged by our custom error handler, we
+		// don't need to log the rest of this request
+		if handled := l.errorHandler(ctx, err, method); handled {
+			return
+		}
+
+		if st.Code() == codes.Internal {
+			ctxlogrus.Extract(ctx).WithError(err).Errorf("internal error response on RPC %s", method)
+			return
+		}
 	}
+
+	ctxlogrus.Extract(ctx).Infof("served RPC %v", method)
+}
+
+// RecoveryMiddleware recovers from panics in the HTTP handler chain, logging
+// an error for Error Reporting
+func RecoveryMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			var err error
+			e := recover()
+			if e == nil {
+				return
+			}
+
+			switch t := e.(type) {
+			case string:
+				err = errors.New(t)
+			case error:
+				err = t
+			default:
+				err = fmt.Errorf("unknown error: %w", t)
```

Comment:
I don't think you can use %w if it's not an error....  also it's explicitly not an error so.... 
```suggestion
				err = fmt.Errorf("unknown panic value: %v", t)
```